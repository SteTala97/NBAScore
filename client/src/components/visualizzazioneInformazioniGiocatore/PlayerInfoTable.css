.playerTable{
  margin: 0 auto;
  margin-top: 35px;
  margin-bottom: 50px;
  }


.player-team-logo {
  height: 125px;
}


.no-info-message{
  text-align: center;
  margin: 0 auto;
  margin-top: 150px;
}