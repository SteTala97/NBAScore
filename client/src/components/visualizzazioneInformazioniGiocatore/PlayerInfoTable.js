import React from 'react';

import TableCell from '@mui/material/TableCell';
import TableRow from '@mui/material/TableRow';
import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableContainer from '@mui/material/TableContainer';
import TableHead from '@mui/material/TableHead';


function isEmptyObject(obj) {
    for (var key in obj) {
      if (Object.prototype.hasOwnProperty.call(obj, key)) {
        return false;
      }
    }
    return true;
}

function calculateAge(date) {
    var ageDifMs = Date.now() - new Date(date);
    var ageDate = new Date(ageDifMs);
    return Math.abs(ageDate.getUTCFullYear() - 1970).toString();
  }


const PlayerInfoTable = (props) => {

    const info = props.playerInfo;
    const team = props.teamInfo;
    
    if (isEmptyObject(info)) {
        return (<div className='no-info-message'>Currently there is no info available about the selected player</div>);
    }
    else {
        return(
            <TableContainer>
                <Table className="playerTable" sx={{ minWidth: 400, maxWidth: 550}}  aria-label="simple table">
                    <TableHead>
                        <TableRow key="1">
                            <TableCell className='title' id="title-table-header" align="center" colSpan={2} style={{fontSize:"x-large"}}>
                                <strong>{info.firstName} {info.lastName}</strong>
                            </TableCell>
                        </TableRow>
                    </TableHead>
                    <TableBody>
                        <TableRow sx={{ '&:last-child td, &:last-child th': { border: 0 } }} >
                            <TableCell id="content-table-info" align="left">
                                🌎 Country: <strong>{info.country === null ? "NA" : info.country}</strong>
                                <br></br>
                                📆 Birth date: <strong>{info.birthday === null ? "NA" : info.birthday}</strong>
                                <br></br>
                                ⏳&nbsp; Age: <strong>{info.birthday === null ? "NA" : calculateAge(info.birthday)}</strong>
                                <br></br>
                                📏 Height: <strong>{info.height === null ? "NA" : info.height.toString().concat('m')}</strong>
                                <br></br>
                                💪🏽 Weight: <strong>{info.weight === null ? "NA" : info.weight.toString().concat('Kg')}</strong>
                                <br></br>
                                #️⃣&nbsp; Jersey number: <strong>{info.jersey === null ? "NA" : info.jersey}</strong>
                                <br></br>
                                ⛹🏽‍♂️&nbsp; Role: <strong>{info.pos === null ? "NA" : info.pos}</strong>
                                <br></br>
                                📅 NBA debut: <strong>{info.nba_start === 0 ? "NA" : info.nba_start}</strong>
                                <br></br>
                                🎓 College: <strong>{info.college === null ? "NA" : info.college}</strong>
                            </TableCell>
                            <TableCell id="content-table-info" className='selectable-team' align="center" onClick={() => {
                                if(!isEmptyObject(team)){
                                    props.setViewteamInfo(true); 
                                }
                            }}
                            >
                                <img src={isEmptyObject(team) ? '' : team.logo === null ? '' : team.logo} className="player-team-logo" alt="Team Logo"></img>
                                <br></br>
                                🏀 Team: <strong>{isEmptyObject(team) ? 'NA' : team.name === null ? "NA" : team.name}</strong>
                            </TableCell>
                        </TableRow>
                        <TableRow>{/* This is an */}
                        </TableRow>{/* empty row */}
                    </TableBody>
                </Table>
            </TableContainer>
        );
    }
};

export default PlayerInfoTable;