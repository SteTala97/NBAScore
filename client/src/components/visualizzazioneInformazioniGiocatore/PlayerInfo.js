import PlayerInfoTable from './PlayerInfoTable';
import TeamInfo from '../visualizzazioneInformazioniSquadra/TeamInfo';
import './PlayerInfoTable.css'
import React, { useEffect, useState } from 'react';


const axios = require('axios');


export default function PlayerInfo (props) {

    // The state is an object containing player's info
    const [playerInfo, setPlayerInfo] = useState("");

    // The state is an object containing the available seasons
    const [season, setSeason] = useState(0);

    // The state is an object containing player's team's info
    const [teamInfo, setTeamInfo] = useState({});
    
    // The state is an object which indicates whether TeamInfo can be viewed
    const [viewteamInfo, setViewteamInfo] = useState(false);

    const playerId = props.playerId;

    useEffect( () => {
        getPlayerInfo(playerId);
        getSeasons();
    }, [playerId]);

    useEffect( () => {
        getTeamInfo(playerId, season)
    }, [season, playerId]);
    

    // ↓ GET calls to the server ↓

    const getPlayerInfo = (playerId) => {
        axios.get('https://nbascore.herokuapp.com/playerInfo', {params: { id: playerId}})
        .then(function (res) {
            setPlayerInfo(res.data[0]); // '[0]' because there's going to be
        })                              //  only one player per id
        .catch(function (error) {
                console.error(error);
        })
    };

    const getSeasons = () => {
        axios.get('https://nbascore.herokuapp.com/seasons')
        .then(function (res) {
            setSeason(res.data.pop()); // consider only the latest season
        })
        .catch(function (error) {
                console.error(error);
        })
    };

    const getTeamInfo = (playerId, season) => {
        if(season) {
            axios.get('https://nbascore.herokuapp.com/playerStats', {params: { id: playerId, season: season }})
            .then(function (res) {
                if(props.setTeamName && res.data.length !== 0) {
                    props.setTeamName(res.data[0].name);
                }
                setTeamInfo(res.data[0]); // '[0]' because there's going to be many elements (player's 
            })                            // statistics per game), but only one is needed to get the team
            .catch(function (error) {
                    console.error(error);
            })
        }
    };

    // ↑ GET calls to the server ↑
    

    return(
        viewteamInfo === true ? 
        <TeamInfo teamName={teamInfo.name}/>
        :
        <React.Fragment>
            <PlayerInfoTable 
                playerInfo = {playerInfo} 
                teamInfo = {teamInfo}
                setViewteamInfo = {props.setTeamClicked ? props.setTeamClicked : setViewteamInfo}
            />
        </React.Fragment>
    );
};