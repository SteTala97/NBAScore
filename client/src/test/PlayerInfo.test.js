import React from "react";

import PlayerInfo from "../components/visualizzazioneInformazioniGiocatore/PlayerInfo";
import PlayerInfoTable from "../components/visualizzazioneInformazioniGiocatore/PlayerInfoTable";

import { shallow } from "enzyme";
import { shallowToJson } from "enzyme-to-json";



const test_player = {
    firstName : "LeBron",
    lastName : "James",
    birthday : "1984-12-30",
    country : "USA",
    nba_start : 2003,
    height : "2.06",
    weight : "113.4",
    college : "St. Vincent-St. Mary HS (OH)",
    jersey : 6,
    pos : "F"
};

const test_team = {
    id : 17,
    name : "Los Angeles Lakers",
    logo : "https://upload.wikimedia.org/wikipedia/commons/thumb/3/3c/Los_Angeles_Lakers_logo.svg/220px-Los_Angeles_Lakers_logo.svg.png"
}



// Test that the components do not raise errors when rendered
describe("Rendering single components", () => {

    it("Renders PlayerInfo component without crashing", () => {
        shallow(<PlayerInfo />);
    });

    it("Renders GameInfoGeneralInfo component without crashing", () => {
        shallow(<PlayerInfoTable playerInfo={test_player} teamInfo = {test_team}/>);
    });
})


// Test that the rendering does not change when the same input data is given
describe("Snapshots testing", () => {
    it("PlayerInfo renders correctly", () => {
        const tree = shallow(<PlayerInfo />);
        expect(shallowToJson(tree)).toMatchSnapshot();
    });

    it("PlayerInfoTable renders correctly when there is no data", () => {
        const tree = shallow(<PlayerInfoTable />);
        expect(shallowToJson(tree)).toMatchSnapshot();
    });
    it("PlayerInfoTable renders correctly when there is data", () => {
        const tree = shallow(<PlayerInfoTable playerInfo={test_player} teamInfo = {test_team}/>);
        expect(shallowToJson(tree)).toMatchSnapshot();
    });
    it("PlayerInfoTable renders correctly when team data is missing", () => {
        const tree = shallow(<PlayerInfoTable playerInfo={test_player}/>);
        expect(shallowToJson(tree)).toMatchSnapshot();
    });
    it("PlayerInfoTable renders correctly when player data is missing", () => {
        const tree = shallow(<PlayerInfoTable teamInfo = {test_team}/>);
        expect(shallowToJson(tree)).toMatchSnapshot();
    });
})
