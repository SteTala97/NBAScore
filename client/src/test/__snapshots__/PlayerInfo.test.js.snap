// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshots testing PlayerInfo renders correctly 1`] = `
<Fragment>
  <PlayerInfoTable
    playerInfo=""
    setViewteamInfo={[Function]}
    teamInfo={Object {}}
  />
</Fragment>
`;

exports[`Snapshots testing PlayerInfoTable renders correctly when player data is missing 1`] = `
<div
  className="no-info-message"
>
  Currently there is no info available about the selected player
</div>
`;

exports[`Snapshots testing PlayerInfoTable renders correctly when team data is missing 1`] = `
<ForwardRef(TableContainer)>
  <ForwardRef(Table)
    aria-label="simple table"
    className="playerTable"
    sx={
      Object {
        "maxWidth": 550,
        "minWidth": 400,
      }
    }
  >
    <ForwardRef(TableHead)>
      <ForwardRef(TableRow)
        key="1"
      >
        <ForwardRef(TableCell)
          align="center"
          className="title"
          colSpan={2}
          id="title-table-header"
          style={
            Object {
              "fontSize": "x-large",
            }
          }
        >
          <strong>
            LeBron
             
            James
          </strong>
        </ForwardRef(TableCell)>
      </ForwardRef(TableRow)>
    </ForwardRef(TableHead)>
    <ForwardRef(TableBody)>
      <ForwardRef(TableRow)
        sx={
          Object {
            "&:last-child td, &:last-child th": Object {
              "border": 0,
            },
          }
        }
      >
        <ForwardRef(TableCell)
          align="left"
          id="content-table-info"
        >
          🌎 Country: 
          <strong>
            USA
          </strong>
          <br />
          📆 Birth date: 
          <strong>
            1984-12-30
          </strong>
          <br />
          ⏳  Age: 
          <strong>
            37
          </strong>
          <br />
          📏 Height: 
          <strong>
            2.06m
          </strong>
          <br />
          💪🏽 Weight: 
          <strong>
            113.4Kg
          </strong>
          <br />
          #️⃣  Jersey number: 
          <strong>
            6
          </strong>
          <br />
          ⛹🏽‍♂️  Role: 
          <strong>
            F
          </strong>
          <br />
          📅 NBA debut: 
          <strong>
            2003
          </strong>
          <br />
          🎓 College: 
          <strong>
            St. Vincent-St. Mary HS (OH)
          </strong>
        </ForwardRef(TableCell)>
        <ForwardRef(TableCell)
          align="center"
          className="selectable-team"
          id="content-table-info"
          onClick={[Function]}
        >
          <img
            alt="Team Logo"
            className="player-team-logo"
            src=""
          />
          <br />
          🏀 Team: 
          <strong>
            NA
          </strong>
        </ForwardRef(TableCell)>
      </ForwardRef(TableRow)>
      <ForwardRef(TableRow) />
    </ForwardRef(TableBody)>
  </ForwardRef(Table)>
</ForwardRef(TableContainer)>
`;

exports[`Snapshots testing PlayerInfoTable renders correctly when there is data 1`] = `
<ForwardRef(TableContainer)>
  <ForwardRef(Table)
    aria-label="simple table"
    className="playerTable"
    sx={
      Object {
        "maxWidth": 550,
        "minWidth": 400,
      }
    }
  >
    <ForwardRef(TableHead)>
      <ForwardRef(TableRow)
        key="1"
      >
        <ForwardRef(TableCell)
          align="center"
          className="title"
          colSpan={2}
          id="title-table-header"
          style={
            Object {
              "fontSize": "x-large",
            }
          }
        >
          <strong>
            LeBron
             
            James
          </strong>
        </ForwardRef(TableCell)>
      </ForwardRef(TableRow)>
    </ForwardRef(TableHead)>
    <ForwardRef(TableBody)>
      <ForwardRef(TableRow)
        sx={
          Object {
            "&:last-child td, &:last-child th": Object {
              "border": 0,
            },
          }
        }
      >
        <ForwardRef(TableCell)
          align="left"
          id="content-table-info"
        >
          🌎 Country: 
          <strong>
            USA
          </strong>
          <br />
          📆 Birth date: 
          <strong>
            1984-12-30
          </strong>
          <br />
          ⏳  Age: 
          <strong>
            37
          </strong>
          <br />
          📏 Height: 
          <strong>
            2.06m
          </strong>
          <br />
          💪🏽 Weight: 
          <strong>
            113.4Kg
          </strong>
          <br />
          #️⃣  Jersey number: 
          <strong>
            6
          </strong>
          <br />
          ⛹🏽‍♂️  Role: 
          <strong>
            F
          </strong>
          <br />
          📅 NBA debut: 
          <strong>
            2003
          </strong>
          <br />
          🎓 College: 
          <strong>
            St. Vincent-St. Mary HS (OH)
          </strong>
        </ForwardRef(TableCell)>
        <ForwardRef(TableCell)
          align="center"
          className="selectable-team"
          id="content-table-info"
          onClick={[Function]}
        >
          <img
            alt="Team Logo"
            className="player-team-logo"
            src="https://upload.wikimedia.org/wikipedia/commons/thumb/3/3c/Los_Angeles_Lakers_logo.svg/220px-Los_Angeles_Lakers_logo.svg.png"
          />
          <br />
          🏀 Team: 
          <strong>
            Los Angeles Lakers
          </strong>
        </ForwardRef(TableCell)>
      </ForwardRef(TableRow)>
      <ForwardRef(TableRow) />
    </ForwardRef(TableBody)>
  </ForwardRef(Table)>
</ForwardRef(TableContainer)>
`;

exports[`Snapshots testing PlayerInfoTable renders correctly when there is no data 1`] = `
<div
  className="no-info-message"
>
  Currently there is no info available about the selected player
</div>
`;
